[stable/jenkins] Support multiple agents

wmcdona89 opened the pull request:
<!--
Thank you for contributing to helm/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/helm/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/helm/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the
history. This will make it easier to identify new changes. The PR will be squashed
anyways when it is merged. Thanks.

For fast feedback, please @-mention maintainers that are listed in the Chart.yaml file.

Please make sure you test your changes before you push them. Once pushed, a CircleCI
will run across your changes and do some initial checks and linting. These checks run
very quickly. Please check the results. We would like these checks to pass before we
even continue reviewing your changes.
-->
#### What this PR does / why we need it:
Add support for easy configuration of additional agents which inherit values from `agent`.

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes helm#20870

#### Special notes for your reviewer:
Moved the pod template JCasC to a new template `jenkins.casc.podTemplate`. `name` fields moved to the top so they stand out. See http://mergely.com/2K5B28FF/?ws=1

Moved the pod template xml configuration to a new template `jenkins.xml.podTemplate`. See http://mergely.com/UNMhX7xx/?ws=1

I chose not to make use of either of the kubernetes plugin casc fields related to agent inheritance.

`defaultsProviderTemplate` - the agent from which all other agents inherit configuration. Merging `agent` into the additional agents in the chart allows all agents to make use of the chart's default configuration for `agent`. To wait for the kubernetes plugin to merge the values at runtime would mean each agent would need to set `sideContainerName` (so container values could be inherited) as well as set any fields which get assigned a default value when nil and thus aren't eligible for inheritance (e.g. `command` and `workingDir`).

`inheritFrom` - indicates from which agent this agent should inherit configuration. I thought an `agent.inheritFrom` field would look out of place in the chart but perhaps this could be added in a future PR.

# `helm template` tests
I placed the following in stable/jenkins/agent-values.yaml
```yaml
agent:
  podName: default
  customJenkinsLabels: default
  # set resources for additional agents to inherit
  resources:
    limits:
      cpu: "1"
      memory: "2048Mi"

# Each additional agent corresponds to a chart `agent` in terms of the configurable values.
additionalAgents:
  maven:
    podName: maven
    customJenkinsLabels: maven
    # An example of overriding the jnlp container
    # sideContainerName: jnlp
    image: jenkins/jnlp-agent-maven
    tag: latest
  python:
    podName: python
    customJenkinsLabels: python
    sideContainerName: python
    image: python
    tag: "3"
    command: "/bin/sh -c"
    args: "cat"
    TTYEnabled: true
```
for helm 2 use `--execute` instead of `--show-only`
## default JCasC with auto-reload (jcasc-config.yaml)
### single agent
```
helm template stable/jenkins \
--show-only templates/jcasc-config.yaml \
--set master.sidecars.configAutoReload.enabled=true \
--set master.JCasC.enabled=true \
--set master.JCasC.defaultConfig=true
```
### multiple agents
```
helm template stable/jenkins \
--show-only templates/jcasc-config.yaml \
--set master.sidecars.configAutoReload.enabled=true \
--set master.JCasC.enabled=true \
--set master.JCasC.defaultConfig=true \
--values stable/jenkins/agent-values.yaml
```
## default JCasC NO auto-reload (config.yaml)
### single agent
```
helm template stable/jenkins \
--show-only templates/config.yaml \
--set master.enableXmlConfig=false \
--set master.sidecars.configAutoReload.enabled=false \
--set master.JCasC.enabled=true \
--set master.JCasC.defaultConfig=true
```
### multiple agents
```
helm template stable/jenkins \
--show-only templates/config.yaml \
--set master.enableXmlConfig=false \
--set master.sidecars.configAutoReload.enabled=false \
--set master.JCasC.enabled=true \
--set master.JCasC.defaultConfig=true \
--values stable/jenkins/agent-values.yaml
```
## XML configuration (config.yaml)
### single agent
```
helm template stable/jenkins \
--show-only templates/config.yaml
```
### multiple agents
```
helm template stable/jenkins \
--show-only templates/config.yaml \
--values stable/jenkins/agent-values.yaml
```

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/mychartname]`)

wmcdona89 committed:
[stable/jenkins] Support multiple agents

Move the kubernetes pod template JCasC to a new template `jenkins.casc.podTemplate`

Update README with an example of using additional agents which inherit
values from `agent` to configure additional kubernetes pod templates

Signed-off-by: Aaron McDonald <wmcdona89@gmail.com>

lgg42 commented on stable/jenkins/README.md:
````diff
@@ -374,6 +377,32 @@ rbac:
   create: true
 ```
 
+### Config as Code with additionl agents
````
I've been testing the updated PR. I think somewhere in the README.md you should state that for this feature to work the following values must set.
```
master.JCasC.enabled: true
master.JCasC.defaultConfig: true
master.JCasC.configScripts: at least one key (cannot be empty {})
```
Specifically the last one. I still don't know why, but if that value is empty the additional agents are not created. It also happened in my PR.

torstenwalter replied:
`master.JCasC.configScripts: at least one key (cannot be empty {})` is not needed anymore

lgg42 replied:
Could please elaborate? I don't get to understand.

torstenwalter replied:
I meant that it's no longer necessary to specify configScripts when enabling JCasC.
https://github.com/helm/charts/issues/21451#issuecomment-601859408

https://github.com/helm/charts/pull/21556 solved that issue.

lgg42 replied:
Thanks for the clarification. I've tried this PR and if I don't set a key for configScripts I don't get the additional agents.
Have you tested this PR with `master.JCasC.configScripts: {}` and got the additional agents?

If affirmative, would you mind sharing your values.yaml files used for the test?

torstenwalter replied:
yes - I tested it with your PR.

torstenwalter commented on stable/jenkins/CHANGELOG.md:
```diff
@@ -5,6 +5,14 @@ numbering uses [semantic versioning](http://semver.org).
 
 NOTE: The change log until version 1.5.7 is auto generated based on git commits. Those include a reference to the git commit to be able to get more details.
 
+## 1.10.2 Support multiple agents in JCasC
```
I think this is worth a minor version increment.

wmcdona89 replied:
Will do.

torstenwalter commented on stable/jenkins/templates/_helpers.tpl:
```diff
@@ -101,47 +101,17 @@ jenkins:
+      {{- end }}
+      {{- /* restore chart agent */}}
+      {{- $_ := set .Values "agent" $agent }}
+      {{- end }}
```
I am impressed by your templating skills. Do you thinks it's best to do the agent inheritence via yaml templating. Isn't it possible to configure it in Jenkins that one agent inherits values from another one. It that's possible then I would prefer it that way as users could decide themselves if they want to inherit values or not.

torstenwalter replied:
We could also set this value so that all templates inherit values:

> defaultsProviderTemplate - The name of the pod template to use for providing default values. When a value is specified and there is a pod template with a matching name, the template will be used as a parent to all other pod templates.

wmcdona89 replied:
> I am impressed by your templating skills.

It was a lot of trial and error :-)

I tried using `defaultsProviderTemplate` but ran into some gotchas as mentioned in the PR notes. Then I discovered `merge` and went that route as I liked how every agent would get the default values the chart provides. Also, I struggled with how to fit the field into the chart since the chart is so "single" agent. But I've since added the `master.JCasC.additionalAgents` field and could add `master.JCasC.defaultAgent`. So I'll revisit `defaultsProviderTemplate` and attempt to document the gotchas for users.

torstenwalter commented on stable/jenkins/templates/_helpers.tpl:
```diff
@@ -159,6 +129,69 @@ unclassified:
+*/}}
+{{- define "jenkins.casc.agent" -}}
+{{- if .Values.agent.enabled -}}
+- name: "{{ .Values.agent.podName }}"
```
If we use Jenkins inheritance for agents then we could simply include yaml below this agent configuration.

wmcdona89 replied:
Sorry. I don't understand what you're suggesting here.

torstenwalter commented on stable/jenkins/values.yaml:
```diff
@@ -443,11 +443,15 @@ agent:
   command:
+  # Default args value is "${computer.jnlpmac} ${computer.name}" but JCasC requires
+  # escaping with ^$ so the value is hardcoded in the templates and applied conditionally.
   args:
```
You are trying to explain these code snippets here, right?

- ./templates/_helpers.tpl

  ```
          {{- if .Values.agent.args }}
          args: "{{ .Values.agent.args }}"
          {{- else }}
          args: "^${computer.jnlpmac} ^${computer.name}"
          {{- end }}
  ```

- ./templates/config.yaml
  ```
  {{- if .Values.agent.args }}
                    <args>{{ .Values.agent.args }}</args>
  {{- else }}
                    <args>${computer.jnlpmac} ${computer.name}</args>
  {{- end }}
  ```

You are absolutely right that this is confusing. If we have an empty value here then we still set a default by using if conditions. 

It would be best if we could set this here to:
```yaml
args: "${computer.jnlpmac} ${computer.name}"
```

and take care of escaping dollar signs in the template.
I moved that part into a separate PR https://github.com/helm/charts/pull/21773

wmcdona89 replied:
Nice! I left a [minor comment](https://github.com/helm/charts/pull/21773/files#r403701427) on the PR.

torstenwalter commented on stable/jenkins/CHANGELOG.md:
```diff
@@ -5,6 +5,17 @@ numbering uses [semantic versioning](http://semver.org).
+- `agent.volumes` added
+- `agent.workingDir` configurable
+- use of envVar as containerEnvVar is deprecated
+
```
There is quire a lot going on in this PR. To make it easier for review I suggest to split it up into smaller ones. e.g. having one PR which just does:

> pod template configuration based on `agent` values moved to a new template `jenkins.casc.podTemplate`

And having the enhancements then in a separate?

wmcdona89 replied:
@torstenwalter I've simplified this PR as much as possible. I'll rebase once https://github.com/helm/charts/pull/21773 is merged.

torstenwalter replied:
It's merged now. Thanks.

wmcdona89 replied:
@torstenwalter I've rebased.

wmcdona89 commented on stable/jenkins/templates/_helpers.tpl:
```diff
@@ -101,50 +101,10 @@ jenkins:
       {{- if .Values.agent.podTemplates }}
         {{- range $key, $val := .Values.agent.podTemplates }}
-          {{- tpl $val $ | nindent 6 }}
+          {{- tpl $val $ | nindent 8 }}
```
bumping to 8 to support folding of `templates` in an editor

torstenwalter commented on stable/jenkins/templates/_helpers.tpl:
```diff
@@ -164,6 +124,53 @@ unclassified:
+    args: "{{ .Values.agent.args }}"
+    {{- else }}
+    args: "^${computer.jnlpmac} ^${computer.name}"
+    {{- end }}
```
`args: "{{ .Values.agent.args | replace "$" "^$" }}"` after rebase.

wmcdona89 replied:
Got it. Thanks for pointing it out.

wmcdona89 commented on stable/jenkins/README.md:
```diff
@@ -516,6 +516,59 @@ agent:
+      {{- end }}
+
+      {{- /* restore .Values.agent */}}
+      {{- $_ := set .Values "agent" $agent }}
```
FYI. In an attempt to provide a layer of abstraction, I tried moving this whole templating block to a helper template and changing the above to simply
`{{- include "jenkins.casc.podTemplates" .Values.additionalAgents }}`

Not surprisingly I encountered scoping issues. See https://github.com/helm/helm/issues/3167
```
{{/*
Returns a list of kubernetes pod templates configured from a list of agent values
*/}}
{{- define "jenkins.casc.podTemplates" -}}
    {{- /* save .Values.agent */}}
    {{- $agent := $.Values.agent }}

    {{- range $name, $additionalAgent := . }}
      {{- /* set .Values.agent to $additionalAgent */}}
      {{- $_ := set $.Values "agent" $additionalAgent }}
      {{- include "jenkins.casc.podTemplate" $ | nindent 0 }}
    {{- end }}

    {{- /* restore .Values.agent */}}
    {{- $_ := set $.Values "agent" $agent }}
{{- end -}}
```
> template: jenkins/templates/jcasc-config.yaml:6:4: executing "jenkins/templates/jcasc-config.yaml" at <include "jenkins.casc.podTemplates" .Values.additionalAgents>: error calling include: template: jenkins/templates/_helpers.tpl:182:19: executing "jenkins.casc.podTemplates" at <$.Values.agent>: nil pointer evaluating interface {}.agent

Commit:
Add first class support for additionalAgents

Signed-off-by: Torsten Walter <torsten.walter@syncier.com>

wmcdona89 committed:
Add additionalAgents in the xml configuration

Move the pod template xml configuration to a new template `jenkins.xml.podTemplate`

Signed-off-by: Aaron McDonald <wmcdona89@gmail.com>

wmcdona89 commented on stable/jenkins/templates/_helpers.tpl:
```diff
@@ -101,46 +101,23 @@ jenkins:
-          {{ tpl .Values.agent.yamlTemplate . | nindent 10 | trim }}
-        yamlMergeStrategy: "override"
+      {{- include "jenkins.casc.podTemplate" . | nindent 8 }}
+    {{- if .Values.additionalAgents }}
```
@torstenwalter I'm assuming you decreased the indent here on purpose to make it more readable.

torstenwalter reviewed (commented):
/lgtm
